Examples/Algorithms/Generators/ActsExamples/Generators/EventGenerator.cpp

```diff
@@ -47,80 +88,483 @@ std::pair<std::size_t, std::size_t> EventGenerator::availableEvents() const {
+    m_outputEvent(ctx, std::move(event));
+  }
+
+  return ProcessCode::SUCCESS;
+}
+
+namespace {
+
+constexpr int kBeamParticleStatus = 4;
+constexpr int kUndecayedParticleStatus = 1;
+
+Acts::Vector4 convertPosition(const HepMC3::FourVector& vec) {
+  return Acts::Vector4(vec.x() * 1_mm, vec.y() * 1_mm, vec.z() * 1_mm,
+                       vec.t() * 1_mm);
+}
+
+std::string printListing(const auto& vertices, const auto& particles) {
+  auto findParticle = [&](SimBarcode particleId) {
+    if (auto it = std::ranges::find_if(particles,
+                                       [&](const auto& particle) {
+                                         return particle.particleId() ==
+                                                particleId;
+                                       });
+        it != particles.end()) {
+      return *it;
+    }
+    throw std::runtime_error("Particle not found");
+  };
+  std::stringstream ss;
+
+  for (const auto& vertex : vertices) {
+    ss << "Vtx:    " << vertex.vertexId() << " at "
+       << vertex.position4.transpose() << "\n";
+    for (const auto& [idx, particleId] : enumerate(vertex.incoming)) {
+      const auto& particle = findParticle(particleId);
+      if (idx == 0) {
+        ss << " I:     ";
+      } else {
+        ss << "        ";
+      }
+      ss << particle;
+      ss << "\n";
+    }
+
+    for (const auto& [idx, particleId] : enumerate(vertex.outgoing)) {
+      const auto& particle = findParticle(particleId);
+      if (idx == 0) {
+        ss << " O:     ";
+      } else {
+        ss << "        ";
+      }
+      ss << particle;
+      ss << "\n";
+    }
+  }
+
+  return ss.str();
+};
+}  // namespace
+
+void EventGenerator::handleVertex(const HepMC3::GenVertex& genVertex,
+                                  SimVertex& vertex,
+                                  std::vector<SimVertex>& vertices,
+                                  std::vector<SimParticle>& particles,
+                                  std::size_t& nSecondaryVertices,
+                                  std::size_t& nParticles,
+                                  std::vector<bool>& seenVertices) {
+  for (const auto& particle : genVertex.particles_out()) {
+    if (particle->end_vertex() != nullptr) {
+      // This particle has an end vertex, we need to handle that vertex
+      const auto& endVertex = *particle->end_vertex();
+      if (seenVertices.at(std::abs(endVertex.id()) - 1)) {
+        // We've seen this vertex before, we don't need to create it again
+        continue;
+      }
+      seenVertices.at(std::abs(endVertex.id()) - 1) = true;
+
+      const auto endVertexPosition = convertPosition(endVertex.position());
+
+      ACTS_VERBOSE("Found secondary vertex at "
+                   << endVertexPosition.transpose());
+      double distance = (endVertexPosition.template head<3>() -
+                         vertex.position4.template head<3>())
+                            .squaredNorm();
```

Comment:
last time I did this I wondered if a Minkowski norm would make more sense but did not conclude on it

Reply:
I guess. I think I'll stick to the spatial distance only. The config parameter is clearly named, so there should not be confusion.